br/pkg/lightning/backend/external/engine.go

```diff
@@ -402,7 +404,28 @@ func (e *Engine) ID() string {
 
 // GetKeyRange implements common.Engine.
 func (e *Engine) GetKeyRange() (startKey []byte, endKey []byte, err error) {
-	return e.startKey, e.endKey, nil
+	return DecodeKeyRangeWithAdapter(e.keyAdapter, e.startKey, e.endKey)
+}
+
+// DecodeKeyRangeWithAdapter decodes start key and end key.
+func DecodeKeyRangeWithAdapter(adapter common.KeyAdapter, start, end []byte) ([]byte, []byte, error) {
+	first, err := adapter.Decode(nil, start)
+	if err != nil {
+		return nil, nil, err
+	}
+	last := end
+	if end[len(end)-1] == 0 {
```

Comment:
I can improve the comments to explain that it's safe to only check the last byte is `0`

Reply:
```
	// when duplicate detection feature is disabled, the end key comes from
	// codec.EncodeBytes or Key.Next(). We can check the the last byte to tell them
	// apart because codec.EncodeBytes will always put non-zero byte at the end.
	//
	// when duplicate detection feature is enabled, the end key comes from
	// DupDetectKeyAdapter.Encode or key.Next(). We try to decode it and check the
	// error.
```

Seems this check is not correct :joy: